File: support_tools/python/upd_commander_checker/src/upd_commander_checker/scanner.py
import ast
from fnmatch import fnmatch
import os
from pathlib import Path

from .classifier import classify_module
from .commander_rules import check_commander
from .container_rules import check_containers
from .data_type_location_rules import check_data_type_locations
from .dependency_rules import check_dependencies
from .flat_layer_rules import check_flat_layers
from .ignore_rules import IgnoreRule, filter_findings, is_path_ignored, load_ignore_rules
from .models import Finding
from .model_attention_rules import (
    ModelGroupOccurrence,
    collect_path_model_group_occurrences,
    model_attention_findings,
)
from .responsibility_rules import check_responsibilities


def scan_path(
    target: Path,
    ignore_patterns: tuple[str, ...] = (),
    upd301_max_inputs: int = 2,
    flat_layer_min_files: int = 12,
    flat_layer_min_direct_percent: int = 80,
    model_group_min_items: int = 3,
    model_group_min_occurrences: int = 2,
) -> list[Finding]:
    root = target if target.is_dir() else target.parent
    classification_root = root if target.is_dir() else None
    try:
        ignore_rules = load_ignore_rules(root)
    except (OSError, UnicodeError) as exc:
        return [Finding(Path(".updcommanderignore"), 1, "UPD001", f"read failed: {exc}")]

    findings: list[Finding] = []
    paths = _python_files(target, root, ignore_patterns, ignore_rules, findings)
    internal_modules = _internal_module_names(paths, root)
    model_occurrences: list[ModelGroupOccurrence] = []
    for path in paths:
        findings.extend(
            _scan_file(
                path,
                root,
                classification_root,
                ignore_rules,
                internal_modules,
                upd301_max_inputs,
            )
        )
        model_occurrences.extend(
            collect_path_model_group_occurrences(
                path,
                root,
                classification_root,
                ignore_rules,
                model_group_min_items,
            )
        )
    findings.extend(
        _filter_location_findings(check_data_type_locations(paths, root), root, ignore_rules)
    )
    findings.extend(model_attention_findings(model_occurrences, model_group_min_occurrences))
    if target.is_dir():
        findings.extend(
            check_flat_layers(
                paths,
                root,
                ignore_rules,
                flat_layer_min_files,
                flat_layer_min_direct_percent,
            )
        )
    return sorted(findings, key=lambda item: (str(item.path), item.line, item.code))


def _python_files(
    target: Path,
    root: Path,
    ignore_patterns: tuple[str, ...],
    ignore_rules: tuple[IgnoreRule, ...],
    findings: list[Finding],
) -> list[Path]:
    if target.is_file():
        paths = [target]
    else:
        paths = []

        def on_error(error: OSError) -> None:
            error_path = Path(error.filename) if error.filename else target
            findings.append(
                Finding(
                    Path(_relative_text(error_path, root)),
                    1,
                    "UPD001",
                    f"read failed: {error}",
                )
            )

        for directory, _, names in os.walk(target, onerror=on_error):
            paths.extend(
                Path(directory) / name
                for name in names
                if name.endswith(".py")
            )

    return [
        path
        for path in paths
        if not _is_ignored(path, root, ignore_patterns, ignore_rules)
    ]


def _internal_module_names(paths: list[Path], root: Path) -> frozenset[str]:
    names: set[str] = set()
    for path in paths:
        names.add(_module_name(path, None))
        names.add(_module_name(path, root))
    return frozenset(name for name in names if name)


def _module_name(path: Path, root: Path | None) -> str:
    module_path = path
    if root is not None:
        try:
            module_path = path.relative_to(root)
        except ValueError:
            pass
    parts = [part for part in module_path.with_suffix("").parts if part != module_path.anchor]
    if parts and parts[-1] == "__init__":
        parts.pop()
    return ".".join(parts)


def _is_ignored(
    path: Path,
    root: Path,
    patterns: tuple[str, ...],
    ignore_rules: tuple[IgnoreRule, ...],
) -> bool:
    relative_text = _relative_text(path, root)
    if any(fnmatch(relative_text, pattern) for pattern in patterns):
        return True
    return is_path_ignored(relative_text, ignore_rules)


def _scan_file(
    path: Path,
    root: Path,
    classification_root: Path | None,
    ignore_rules: tuple[IgnoreRule, ...],
    internal_modules: frozenset[str],
    upd301_max_inputs: int,
) -> list[Finding]:
    try:
        source = path.read_text(encoding="utf-8")
        tree = ast.parse(source, filename=str(path))
    except (OSError, UnicodeError) as exc:
        return [Finding(path, 1, "UPD001", f"read failed: {exc}")]
    except SyntaxError as exc:
        return [Finding(path, exc.lineno or 1, "UPD002", f"syntax: {exc.msg}")]

    module = classify_module(path, classification_root)
    findings = check_dependencies(tree, module, internal_modules)
    findings.extend(check_commander(tree, module))
    findings.extend(check_containers(tree, module, upd301_max_inputs))
    findings.extend(check_responsibilities(tree, module))
    return filter_findings(findings, source, _relative_text(path, root), ignore_rules)


def _filter_location_findings(
    findings: list[Finding],
    root: Path,
    ignore_rules: tuple[IgnoreRule, ...],
) -> list[Finding]:
    filtered: list[Finding] = []
    for finding in findings:
        source_path = root / finding.path
        try:
            source = source_path.read_text(encoding="utf-8")
        except (OSError, UnicodeError):
            continue
        filtered.extend(
            filter_findings([finding], source, finding.path.as_posix(), ignore_rules)
        )
    return filtered


def _relative_text(path: Path, root: Path) -> str:
    try:
        return path.relative_to(root).as_posix()
    except ValueError:
        return path.as_posix()

File: support_tools/go/upd_commander_checker/internal/checker/scanner.go
package checker

import (
	"bufio"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

func ScanPath(target string, cliIgnore []string) []Finding {
	return ScanPathWithAllThresholds(target, cliIgnore, 2, 12, 80, 3, 2)
}

func ScanPathWithUpd301MaxInputs(target string, cliIgnore []string, upd301MaxInputs int) []Finding {
	return ScanPathWithAllThresholds(target, cliIgnore, upd301MaxInputs, 12, 80, 3, 2)
}

func ScanPathWithThresholds(target string, cliIgnore []string, upd301MaxInputs int, flatLayerMinFiles int, flatLayerMinDirectPercent int) []Finding {
	return ScanPathWithAllThresholds(target, cliIgnore, upd301MaxInputs, flatLayerMinFiles, flatLayerMinDirectPercent, 3, 2)
}

func ScanPathWithAllThresholds(target string, cliIgnore []string, upd301MaxInputs int, flatLayerMinFiles int, flatLayerMinDirectPercent int, modelGroupMinItems int, modelGroupMinOccurrences int) []Finding {
	return ScanPathWithCommonRoots(target, cliIgnore, upd301MaxInputs, flatLayerMinFiles, flatLayerMinDirectPercent, modelGroupMinItems, modelGroupMinOccurrences, []string{"common", "shared"})
}

func ScanPathWithCommonRoots(target string, cliIgnore []string, upd301MaxInputs int, flatLayerMinFiles int, flatLayerMinDirectPercent int, modelGroupMinItems int, modelGroupMinOccurrences int, commonRoots []string) []Finding {
	root := target
	info, err := os.Stat(target)
	targetIsFile := err == nil && !info.IsDir()
	if targetIsFile {
		root = filepath.Dir(target)
	}
	contextRoot := root
	if targetIsFile {
		contextRoot = singleFileContextRoot(target)
	}

	rules, ignoreErr := loadIgnoreRules(root)
	if ignoreErr != nil {
		return []Finding{{Path: ".updcommanderignore", Line: 1, Code: "UPD001", Message: "read failed: " + ignoreErr.Error(), Severity: "error"}}
	}
	var findings []Finding
	var paths []string
	walkErr := filepath.WalkDir(target, func(path string, entry os.DirEntry, walkErr error) error {
		if walkErr != nil {
			rel, relErr := filepath.Rel(root, path)
			if relErr != nil {
				rel = path
			}
			findings = append(findings, Finding{Path: filepath.ToSlash(rel), Line: 1, Code: "UPD001", Message: "read failed: " + walkErr.Error(), Severity: "error"})
			return nil
		}
		if entry == nil || entry.IsDir() || filepath.Ext(path) != ".go" {
			return nil
		}
		rel, relErr := filepath.Rel(root, path)
		if relErr != nil {
			rel = path
		}
		relText := filepath.ToSlash(rel)
		if pathIgnored(relText, cliIgnore) || IsPathIgnored(relText, rules) {
			return nil
		}
		paths = append(paths, path)
		return nil
	})
	if walkErr != nil {
		findings = append(findings, Finding{Path: filepath.ToSlash(target), Line: 1, Code: "UPD001", Message: "read failed: " + walkErr.Error(), Severity: "error"})
	}

	dependencyPaths := paths
	if targetIsFile {
		dependencyPaths = dependencyContextPaths(contextRoot)
	}
	internalPackages := internalPackagePaths(dependencyPaths, contextRoot)
	modelOccurrences := []ModelGroupOccurrence{}
	for _, path := range paths {
		rel, relErr := filepath.Rel(root, path)
		if relErr != nil {
			rel = path
		}
		classificationRel, classificationErr := filepath.Rel(contextRoot, path)
		if classificationErr != nil {
			classificationRel = path
		}
		relativeText := filepath.ToSlash(rel)
		classificationText := filepath.ToSlash(classificationRel)
		findings = append(
			findings,
			scanFile(path, relativeText, classificationText, rules, internalPackages, upd301MaxInputs)...,
		)
		modelOccurrences = append(
			modelOccurrences,
			collectModelGroupOccurrences(path, relativeText, classificationText, rules, modelGroupMinItems)...,
		)
	}
	findings = append(findings, checkDataTypeLocations(paths, root, rules)...)
	findings = append(findings, checkCommonUsage(paths, root, commonRoots, rules)...)
	findings = append(findings, modelAttentionFindings(modelOccurrences, modelGroupMinOccurrences)...)
	if !targetIsFile {
		findings = append(findings, checkFlatLayers(paths, root, rules, flatLayerMinFiles, flatLayerMinDirectPercent)...)
	}
	sort.Slice(findings, func(i, j int) bool {
		if findings[i].Path != findings[j].Path {
			return findings[i].Path < findings[j].Path
		}
		if findings[i].Line != findings[j].Line {
			return findings[i].Line < findings[j].Line
		}
		return findings[i].Code < findings[j].Code
	})
	return findings
}

func scanFile(path string, rel string, classificationPath string, rules []IgnoreRule, internalPackages []string, upd301MaxInputs int) []Finding {
	fset := token.NewFileSet()
	file, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
	if err != nil {
		return []Finding{{Path: rel, Line: 1, Code: "UPD002", Message: "syntax error", Severity: "error"}}
	}
	lines := readLines(path)
	source := ClassifyPath(classificationPath)
	var findings []Finding
	findings = append(findings, checkDependencies(file, fset, source, rel, lines, rules, internalPackages)...)
	findings = append(findings, checkCommander(file, fset, source, rel, lines, rules)...)
	findings = append(findings, checkContainerBoundaries(file, fset, source, rel, lines, rules, upd301MaxInputs)...)
	findings = append(findings, checkResponsibilities(file, fset, rel, lines, rules)...)
	return findings
}

func readLines(path string) []string {
	file, err := os.Open(path)
	if err != nil {
		return nil
	}
	defer file.Close()
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

func pathIgnored(path string, patterns []string) bool {
	for _, pattern := range patterns {
		if globMatch(path, pattern) {
			return true
		}
	}
	return false
}

func globMatch(path string, pattern string) bool {
	pattern = filepath.ToSlash(pattern)
	if strings.HasSuffix(pattern, "/**") {
		return strings.HasPrefix(path, strings.TrimSuffix(pattern, "**"))
	}
	matched, _ := filepath.Match(pattern, path)
	return matched
}

var flatLayerExcludedDirs = map[string]bool{
	"generated":   true,
	"third_party": true,
	"vendor":      true,
	"external":    true,
	"build":       true,
}

type flatLayerCount struct {
	total  int
	direct int
}

func checkFlatLayers(paths []string, root string, rules []IgnoreRule, minFiles int, minDirectPercent int) []Finding {
	counts := map[string]flatLayerCount{}
	for _, path := range paths {
		relative, err := filepath.Rel(root, path)
		if err != nil {
			continue
		}
		relative = filepath.ToSlash(relative)
		parts := strings.Split(relative, "/")
		if hasExcludedFlatLayerDirectory(parts) {
			continue
		}
		layerIndex := flatLayerRootIndex(parts)
		if layerIndex < 0 {
			continue
		}
		layerRoot := strings.Join(parts[:layerIndex+1], "/")
		count := counts[layerRoot]
		count.total++
		if len(parts) == layerIndex+2 {
			count.direct++
		}
		counts[layerRoot] = count
	}

	findings := []Finding{}
	for layerRoot, count := range counts {
		if count.total < minFiles || count.direct*100 < count.total*minDirectPercent {
			continue
		}
		if IsIgnored(layerRoot, "UPD405", "", rules) {
			continue
		}
		findings = append(findings, Finding{
			Path:     layerRoot,
			Line:     1,
			Code:     "UPD405",
			Message:  "large flat layer reduces navigability; consider grouping related responsibilities",
			Severity: "attention",
		})
	}
	return findings
}

func hasExcludedFlatLayerDirectory(parts []string) bool {
	for _, part := range parts[:max(0, len(parts)-1)] {
		if flatLayerExcludedDirs[strings.ToLower(part)] {
			return true
		}
	}
	return false
}

func flatLayerRootIndex(parts []string) int {
	scopeStart := 0
	for index := 0; index+2 < len(parts); index++ {
		if appRoots[strings.ToLower(parts[index])] {
			scopeStart = index + 2
		}
	}
	layerIndex := -1
	for index := scopeStart; index+1 < len(parts); index++ {
		if layerNames[strings.ToLower(parts[index])] {
			layerIndex = index
		}
	}
	return layerIndex
}
